### ISSUE TITLE
Permissions checking

### ISSUE THREAD

**Issue by peternied** [opensearch-project/OpenSearch#5859]
### Dependency
Requires `Permissions granting' tasks

### Use case:
`Subject.checkPermission(...)` uses stored permissions grants
```java
// Example usage
final Subject currentSubject = Identity.getAuthManager().getSubject();
final UnauthorizedException unauthorizedException = currentSubject.checkPermission(...);
```

### Acceptance Criteria:
- Reject invalid permissions checks
- Add debug/trace logging for checks

**Comment by scrawfor99** [comment]
@opensearch-project/security do we think want the permission checking to check for all permissions that a subject has or just a target permission? I also do not know what we would consider an invalid permission check or whether this is best served as a method of the InternalSubjects or as a function in the PermissionHandler class which then takes in a subject as an argument. 

Personally I am leaning towards the latter as I like the flow of using handlers to address the assignment and moving or authn/authz. That would let you override the handler if you were ever to want to make changes to the operation of the function calls without redefining what a subject is etc. I am open to thoughts however.

**Comment by peternied** [comment]
Just a target permission.  While there could be a way to query the permissions storage system, this might be "expensive" and shouldn't be the default.

I omitted the exact value so you could design a class that fits the granting model better.   I've updated the acceptance criteria with this task.

**Comment by cwperks** [comment]
Capturing some pseudo code that @scrawfor99 and I just discussed. This pseudo-code relies on permission naming scheme suggested here: https://github.com/opensearch-project/OpenSearch/issues/5889#issuecomment-1402729760

The comparing action pattern to requested pattern can be trivially accomplished using shiro's `WildcardPermission`

When determining if the resource matches the pattern we need a way to resolve a list of index patterns to concrete indices.

Then the question becomes `Are all requested concrete indices in the set of concrete indices in the permission`?

example:

```
OpenSearchPermission permission1 = new OpenSearchPermission("indices:data/read/*|my-index1");
permission1.implies("indices:data/read/search|my-index1");
```

OpenSearchPermission outline:

```
// Permission interface is from shiro
public class OpenSearchPermission implements Permission {

    @Override
    public boolean implies(Permission p) {
        // pseudo-code

        // Examples
        // OpenSearch permission instance = new OpenSearchPermission("indices:data/read/search|my-index*,other-index*")

        // caller is asking does "indices:data/read/search|my-index*,other-index*"
        // imply "indices:data/read/search|my-index1"
        // YES

        // caller is asking does "indices:admin/create|my-index1"
        // imply "indices:data/read/search|my-index1"
        // NO

        // caller is asking does "indices:data/read/*|my-index1"
        // imply "indices:data/read/search|my-index1"
        // YES

        // OpenSearch Permission: indices:data:read:* - replace forward slash with color to adhere to shiro's convention
        // Requested Permission: indices:data:read:search

        // WildcardPermission wp = new WildcardPermission(actionNameFromPermission.replace("/", ':'));
        // if (!wp.implies(new WildcardPermission(requestedActionName.place("/", ":")))) {
        //     return false;
        // }

        // Next question: my_index1 in my_index*,other_index*
        // Resolve my_index*,other_index* to concrete indices
        // Then is my_index1 in resolved indices

        // Cluster has my_index1, my_index2, other_index1

        // resolveIndices("my-index*,other-index*) -> List.of("my-index1", "my-index2", "my-index3")

        // String actionName = extractActionName(p.toString());
        // List<String> indexPatterns = extractIndexPatterns(p.toString());
        // List<String> resolvedIndices = resolveIndexPatterns(indexPatterns);
        return false;
    }
}
```

**Comment by scrawfor99** [comment]
> It may make sense to modify the `ActionRequest` abstract class to add a new default method.
> 
> ```
> public enum ParamType {
>     INDEX_PATTERN("index-pattern")
> }
> 
> public AuthorizationInfo getAuthorizationParams(); // TODO find a good name
> 
> public class SearchRequest extends ActionRequest {
>     ...
> 
>     @Override
>     public AuthorizationInfo getAuthorizationParams() {
>       // AuthorizationInfo may be a data-structure like
>       // {
>       //   "required": [ParamType.INDEX_PATTERN],
>       //   "optional": [] // empty
>       //   "values": {
>       //     "index-pattern": ["my-index*, other-index*"]
>       //   }
>       // }
>     }
> }
> ```
> 
> This can then be used inside the `SecurityFilter` (where privilege evaluation is performed) to determine which action require resource patterns and to extract the values from the request so that privilege evaluation can be performed. It would be nice if we could obtain this information from the request itself, but I'm not sure if this will work with some requests like `BulkRequest` and `MultiGetRequest`.
> 
> In the security plugin it is assumed that actions beginning with `indices` require an index pattern when evaluation privileges (with few exceptions like `index_template` and `template`). When extracting indices from an `ActionRequest` the security plugin uses a class called `IndexResolverReplacer` and details can be found here: https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java#L529-L711

@cwperks that could make sense.

How do you think this would connect with the `implies()` method that was being added to the `OpenSearchPermission` class as part of extending Shiro Permissions? It seems like this would mean that under this we would have the SecurityFilter determining the allow/disallow versus the `implies()`? I know that is how things are currently handled but I had thought we were planning to restructure things around the `implies()` functionality and basically just making the requests cause a search of all Subject permissions looking for any that would imply they had the needed permission? Right now all the information is being stored in the OpenSearchPermission object and then being evaluated with `implies()`. 

Under this suggestion it seems like the evaluation would be based in a version of the SecurityFilter.

**Comment by cwperks** [comment]
@scrawfor99 There's 2 problems I am trying to think of a solution for:

1. How to determine which params are required for authorization. i.e. indices required for a SearchAction/SearchRequest
2. How to extract the params from the ActionRequest

By having a convenient method for 1 & 2, it will make the logic inside of the Filter that performs Authorization (currently named `SecurityFIlter`) simpler and IMO it could be easier to read the code to determine how requests are authorized.

Some pseudo-code for the `SecurityFilter` could be - and yes we still use `implies`, this occurs upstream of the `implies` call. 

```
Subject currentSubject = Identity.getAuthManager().getSubject();

// For SearchRequest could be a data-structure like
// {
//   "index-pattern": ["my-index*", "other-index*"]
// }
Map<String, List<String>> authParams = request.getAuthorizationParams();

// QueryParams.toString(...) transforms the map into "index-pattern=my-index*&index-pattern=other-index*" 
String permissionRequest = actionName + "?" QueryParams.toString(authParams); 

if (log.isDebugEnabled()) {
    log.debug("Current Subject: {}", currentSubject);
    // isPermitted calls implies under the hood
    log.debug("Current Subject is permitted to perform {}: {}", permission, currentSubject.isPermitted(permission));
}
```

**Comment by cwperks** [comment]
Also thinking outside of only `index-pattern` and towards extensions, it may be possible to create a framework for security that extensions developers can utilize following `index-patterns` as an example. 

Not sure if this makes sense, but consider the example:

```
Bob is granted permission to edit anomaly detectors created by Alice
```

Can we craft a permission grant to look something like:

```
Bob has extensions:ad/detectors/edit?username=alice123
```

And in this instance when evaluating whether Bob has permission to edit Alice's detector it uses `username` instead of `index-pattern`?

**Comment by peternied** [comment]
>Bob is granted permission to edit anomaly detectors created by Alice

We should not support permissions like this.

Only values that are included in the request should be permissionable - this has a huge implication on caching.  Since the creator `Alice` isn't included in the query `GET /plugins/ad/get_detector/1234` we cannot cache and effectively have to retrieve the object before we can know if the user sees it.

**Comment by cwperks** [comment]
@peternied I'll try to clarify more in a following comment with better thought out details, but in short I was thinking about the possibility of extending the index-pattern checking to be useful in more scenarios. Index pattern permissioning is fairly well understood.

- User is granted search permission on `logs-*`
- User is able to search for indices like `logs-2023-01-03`, `logs-2023-01-04`, etc.

What if extensions can leverage the glob matching for other use-cases?

i.e. Let's take Anomaly Detection as an example. Let's say that AD can be modified to be able to store detectors in groups.

1. my-group
   - detector1
   - detector2
   - detector3

2. other-group
   - detector4
   - detector5
   - detector6

Would it be possible to design the permissions to give some useful conventions for extensions developers? Since glob matching is an understood convention, could it be extended for other uses?

**Comment by scrawfor99** [comment]
@peternied, to follow up on your response to Craig, I was wondering why we could not include the creator into the request? Could we not make that an option in the REST request? Is it because the REST APIs for the different plugins are already written (for instance [anomaly detection](https://opensearch.org/docs/latest/monitoring-plugins/ad/api/))? Maybe we could have it be a field in the API but use Identity to resolve the field with the user in the name? I believe we run the request through security anyways so we could always strip that claim from the object before passing it on to be resolved by the plugin. 

That being said, this seems a bit complicated for starting things out. Maybe we can start with some of the simpler options and once we get a base go from there. 

@cwperks I do think that your idea about the groupings could be useful. However, I don't know if that would be better handled on the plugin's side of things. A plugin could add it to their API to resolve the `group=my-group` aspect into the different detectors. If we tried to have it be part of the Identity REST API we would need to know all the structures in the different plugins and extensions since I don't know how else we would be able to guarantee they all use the same naming conventions.

**Comment by cwperks** [comment]
@scrawfor99 @peternied The idea is not user or groups in particular, it's to create a framework that extensions can use to leverage built in conventions to implement security more easily. Extensions may still roll their own security for extension constructs, but it may be useful to have some conventions built-in to bootstrap extensions creators. Since index-patterns are known within OpenSearch, I was proposing to make the same pattern more generic to be useful elsewhere.

From the comment earlier this morning (https://github.com/opensearch-project/OpenSearch/issues/5859#issuecomment-1408885935), I raised the idea of having a method on an ActionRequest to return params required for authorization to try to solve a couple of problems:

1. How to determine which params are required for authorization. i.e. indices required for a SearchAction/SearchRequest
2. How to extract the params required for authorization from the ActionRequest

Let's see if I can illustrate the problem in pseudo-code:

```
public boolean canPerform(String action, ActionRequest request) { // This is a placeholder for SecurityFilter.apply0

   Subject currentSubject = Identity.getAuthManager().getSubject();
  
    // Current subject resolves to list of permissions ["cluster:monitor/health", "indices:data/read/search?index-pattern=my-index*&index-pattern=other-index*"]

   // Scenario 1
   currentSubject.isPermitted("cluster:monitor/health") // TRUE - This action does not require any params

   // Scenario 2
   currentSubject.isPermitted("indices:data/read/search?index-pattern=my-index1") // True = This matches the second permission granted in the list
   
   // For Scenario 2 how does the permissioning system know that index-pattern is required and how does it extract them from the ActionRequest?
}
```

**Comment by peternied** [comment]
> That being said, this seems a bit complicated for starting things out. Maybe we can start with some of the simpler options and once we get a base go from there.

I really like this approach, and that is a good way to circle back on this topic later.

> Could a user be granted a permission like: `extensions:ad/detector/edit?group=my-group`

That permission pattern requires indirect knowledge of the resource.  We would have to dynamically compute those indirect knowledge on every permission look up.  If the data you are trying to check permission for `GET /plugins/ad/get_detector/1234` on is spread over 30 nodes in a cluster, you will need to get a response from every node before you can evaulate they have access or not to that resource.

If the pattern is included in the access pattern, e.g. lets consider a virtual file system extension that lets you interact with a file system, yes this is silly, `GET /plugin/virtual_file_system/home/petern/.bashrc` I think we could author permissions grants around things under `home`, files named `.bashrc`, or explict `/home/petern/.bashrc` path 

- `extensions:vfs/readfile|startsWith=home`
- `extension:vfs.readfile|endsWith=.bashrc`
- `extension:vfs.readfile|path=home/petern/.bashrc`

Those permissions are viable, because all associated parameters are included in the request.

**Comment by peternied** [comment]
> From the comment earlier this morning (https://github.com/opensearch-project/OpenSearch/issues/5859#issuecomment-1408885935), I raised the idea of having a method on an ActionRequest to return params required for authorization to try to solve a couple of problems...

@cwperks If we are making one-way door decisions we should be careful, but augmenting permissions language seems like a good two-way door opportunity.  Lets engage a use case from a plugin we are attempting to migrate them.  Can we start with a simpler/smaller design and iterate on if for scenarios from extensions?

**Comment by cwperks** [comment]
Of course, we need to walk before we can run. The authorization system being built out needs to support all native actions and actions from default plugins so we have a lot to work on there. 

This PR provides a template which can be iterated on: https://github.com/opensearch-project/OpenSearch/pull/6029

In the near future to continue iterating on permissions:

1. Expand to more actions including actions like BulkRequest, MultiGetRequest, SnapshotRestoreRequest and others.
2. Expand test coverage to cover as many scenarios as possible. If there are good tests in the security plugin that capture authz then those should be ported over.

**Comment by scrawfor99** [comment]
@cwperks Thank you for your detailed replies. I definitely see where you are coming from and don't want to seem discouraging. I think your idea may be a great option long term, I just wanted to suggest that we start with one or two options and see how that may work and go from there. I think that we could run into more and more complexity with the implementation and design the more aspects considered. I will keep an eye on the linked PR you shared and do my best to build alongside you. I think that you have a good understanding of the requirements we need to design around so I am glad to see you voice your opinion.

**Comment by scrawfor99** [comment]
@cwperks that makes sense. I agree about all the aspects of a good design as well. What are your thoughts on something like what I started over [here](https://github.com/opensearch-project/OpenSearch/pull/6110/files). I took your permission requirements idea and thought about making an enum that could be queried. I don't want to write the whole thing if that is not something you think could work though. 

I was thinking this would let you query the enum for the requirements similar to what you had mentioned but keeping it separate from the request itself. I don't know if that is worthwhile. I am also curious for your thoughts @peternied.

**Comment by cwperks** [comment]
@scrawfor99 To keep it simpler to start with I suggest we assume that `indices:` permissions require an index patterns and other prefixes do not. We know the list of exceptions and can handle them accordingly. Once the permissions checking covers native actions and default plug-in actions then we can work backwards from extensions use-cases to see if the permissions checking system can be extended to provide additional benefits to extensions.

**Comment by scrawfor99** [comment]
> @scrawfor99 To keep it simpler to start with I suggest we assume that `indices:` permissions require index patterns and other prefixes do not. We know the list of exceptions and can handle them accordingly. Once the permissions checking covers native actions and default plug-in actions then we can work backwards from extensions use-cases to see if the permissions checking system can be extended to provide additional benefits to extensions.

Thank you for your reply @cwperks. I am a little confused I think. I thought that what you have linked over in #6029 was intended to be able to fetch the required parameters for all of the permissions. This is just an alternative way of storing the permission requirements based on the request type which does not require modifying every request to be able to provide that info. So it is the same idea as what you created but just stored differently. To that point, I am definitely not sure that such a complicated enum is the way to go, but I am not sure I quite understand about this vs. #6029. 

If we are just going to assume that only `indices` permissions require index patterns and nothing else does, then why do we need any logic before anyways? We don't need to fetch the parameters because we will just have a general rule. Could you clarify this for me?

**Comment by cwperks** [comment]
@scrawfor99 I'll try to point out what I see as the problems with more examples.

**1. How to determine which params are required for authorization. i.e. indices required for a SearchAction/SearchRequest**

- This is needed during permission granting to be able to validate the request and ensure the user has supplied a valid permission to be added to a user. A user trying to grant `indices:data/read/search` without an index pattern should receive an error message with a clear reason why the grant was invalid.
- There currently is no systematic way to do this so I am proposing, for the time being, that `indices:` actions are assumed to require index-pattern(s) except for the handful of known exceptions.

**2. How to extract the params from the ActionRequest**

- Take a look at the current form of `SecurityFilter`: https://github.com/opensearch-project/OpenSearch/blob/feature/identity/sandbox/modules/identity/src/main/java/org/opensearch/identity/SecurityFilter.java#L77-L85

```
Subject currentSubject = Identity.getAuthManager().getSubject();
// can be multiple index patterns
Optional<String[]> resources = Optional.empty();
if (request instanceof IndicesRequest) {
    String[] indices = ((IndicesRequest) request).indices();
    if (indices != null) {
        resources = Optional.of(((IndicesRequest) request).indices());
    }
}
```

This is not a systematic way of extracting the index pattern(s) from all request types. There needs to be a systematic way of extracting index-pattern(s) - or more generally required params for authorization - to have a convenient way of forming the permission request that will go into `currentSubject.isPermitted(permissionRequested);`

-----

I'm not sure if this is the best place for discussing all of the implementation details. I will create issues for the above 2 problems to discuss in greater depth.

**Comment by scrawfor99** [comment]
This is resolved with #6220 which allows for a REST api to fetch all permissions associated with a given principal. A nonexistant principal being checked with throw an error and since it does not target a specific permission but instead checks for all permissions granted to a principal, we do not need to worry about the formatting of the permissions. There is a logger in the class, but it is not currently being used as it is not clear what would need to be logged outside of the exceptions. This can be easily changed in a follow-up PR. I feel this issue has been addressed.
